mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala

```diff
@@ -0,0 +1,472 @@
+    new DistributedLDAModel(state)
+  }
+}
+
+
+object LDA {
+
+  /*
+    DEVELOPERS NOTE:
+
+    This implementation uses GraphX, where the graph is bipartite with 2 types of vertices:
+     - Document vertices
+        - indexed {0, 1, ..., numDocuments-1}
+        - Store vectors of length k (# topics).
+     - Term vertices
+        - indexed {-1, -2, ..., -vocabSize}
+        - Store vectors of length k (# topics).
+     - Edges correspond to terms appearing in documents.
+        - Edges are directed Document -> Term.
+        - Edges are partitioned by documents.
+
+    Info on EM implementation.
+     - We follow Section 2.2 from Asuncion et al., 2009.  We use some of their notation.
+     - In this implementation, there is one edge for every unique term appearing in a document,
+       i.e., for every unique (document, term) pair.
+     - Notation:
+        - N_{wkj} = count of tokens of term w currently assigned to topic k in document j
+        - N_{*} where * is missing a subscript w/k/j is the count summed over missing subscript(s)
+        - gamma_{wjk} = P(z_i = k | x_i = w, d_i = j),
+          the probability of term x_i in document d_i having topic z_i.
+     - Data graph
+        - Document vertices store N_{kj}
+        - Term vertices store N_{wk}
+        - Edges store N_{wj}.
+        - Global data N_k
+     - Algorithm
+        - Initial state:
+           - Document and term vertices store random counts N_{wk}, N_{kj}.
+        - E-step: For each (document,term) pair i, compute P(z_i | x_i, d_i).
+           - Aggregate N_k from term vertices.
+           - Compute gamma_{wjk} for each possible topic k, from each triplet.
+             using inputs N_{wk}, N_{kj}, N_k.
+        - M-step: Compute sufficient statistics for hidden parameters phi and theta
+          (counts N_{wk}, N_{kj}, N_k).
+           - Document update:
+              - N_{kj} <- sum_w N_{wj} gamma_{wjk}
+              - N_j <- sum_k N_{kj}  (only needed to output predictions)
+           - Term update:
+              - N_{wk} <- sum_j N_{wj} gamma_{wjk}
+              - N_k <- sum_w N_{wk}
+   */
+
+  /**
+   * :: DeveloperApi ::
+   *
+   * Document with an ID.
+   *
+   * @param counts  Vector of term (word) counts in the document.
+   *                This is the "bag of words" representation.
+   * @param id  Unique ID associated with this document.
+   *            Documents should be indexed {0, 1, ..., numDocuments-1}.
+   *
+   * TODO: Can we remove the id and still be able to zip predicted topics with the Documents?
+   *
+   * NOTE: This is currently marked DeveloperApi since it is under active development and may
+   *       undergo API changes.
+   */
+  @DeveloperApi
+  case class Document(counts: Vector, id: Long)
+
+  /**
+   * Vector over topics (length k) of token counts.
+   * The meaning of these counts can vary, and it may or may not be normalized to be a distribution.
+   */
+  private[clustering] type TopicCounts = BDV[Double]
```

Comment:
Is there a reason to have this as a Double rather than an Int?
+ Makes normalizing easier.
- Uncertain what signals what these counts mean (aka, whether its in a normalized state or not)
- Affects the Gibbs implementation since its part of the LDA API, which currently uses Array[Int].

Computationally, normalizing should cost about as much as returning a new array from an array (or vector) of integer counts. Thoughts?